Pr kronsubmodule

hbrunie opened the pull request:
## Proposed changes

Change kronmult external library as a submodule of Asgard.

## What type(s) of changes does this code introduce?

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?
Cori GPU
## Checklist

- [x ] this PR is up to date with current the current state of 'develop'
- [ ] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

hbrunie commented:
The tree of commit in the PR is a bit of a mess... As the PR should be pretty simple, I suggest we **squash** and merge them all into just one commit.

bmcdanie commented:
@hbrunie that sounds good to me.

hbrunie committed:
Change external dependance (kron).

kronmult to become a submodule.

bmcdanie commented on CMakeLists.txt:
```diff
@@ -105,14 +105,16 @@ if(ASGARD_USE_CUDA)
 endif()
 
 # build component to interface with Ed's kronmult lib
+##TODO: insert the interface directly inside kronmult dir
```
can you help me understand this comment?

hbrunie replied:
This comment is very unclear, and we should probably delete it, or rewrite it.
I am not enough familiar with CMAKE to know the best way to implement the link between kronmult and asgard. So I just made something that works. But I think there is another way to implement this using CMAKE interface feature.

bmcdanie replied:
I think this way is good for me - we do have a `PRIVATE` kronmult_cuda dependency on kron

bmcdanie replied:
I've only used the `INTERFACE` to link something that needs to export symbols, but not consume them itself. you may mean something else though - I'm thinking you're referring to linking as `INTERFACE` rather than `PRIVATE`

bmcdanie commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "contrib/kronmult"]
+	path = contrib/kronmult
+	url = https://github.com/project-asgard/kronmult.git
```
is this always going to pull the latest commit of master for kronmult?

hbrunie replied:
My understanding is that it will not. 
To update submodule I found this commands: 
```
git pull --recurse-submodules
git submodule update --remote --recursive
```
There are other ways.

bmcdanie replied:
ok - I do not want that behavior (automatically pulling latest), because I don't want development changes there to implicitly ripple into ASGarD.

rather, I'd prefer we set the gitlink to the latest release. I'm not super familiar with that process, but looks like it can be done this way? https://stackoverflow.com/questions/43671101/how-to-add-a-specific-release-of-a-repo-as-a-submodule

bmcdanie replied:
after reading up on this, I see the gitlink will be fixed to a particular commit in the submodule repo unless we alter it, so I am good with this.

bmcdanie commented on contrib/contrib.cmake:
```diff
@@ -184,37 +184,36 @@ endif ()
+    # also install it
+    install(DIRECTORY ${KRON_INCLUDE_DIR}/kronmult DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+
+    # for convenience setup a target
```
also, can you help me understand these commands? thanks.

hbrunie replied:
That is linked to the comment I made above, you made a comment on this.
I don't understand exactly how INTERFACE works in CMAKE, but I understood it was one way of dealing with dependences. 
I built the kronmult_interface (CMAKE INTERFACE) to be able to use it in the asgard CMakeLists.txt directly instead of linking "by hand" the library. But I gave up to finish it properly.

So: I can either continue reading on it and trying to make it using CMAKE INTERFACE (changing Asgard CMakeLists.txt) or we can delete these lines.

bmcdanie replied:
ah I can help some with this (as far as I know)

`INTERFACE` flag when linking means the target does not itself need the symbols linked in, but things that link to the target do.

in our case, we need `PRIVATE`, because `kronmult_cuda` does need the symbols defined in `kron`. if `kronmult_cuda` did not need the symbols, but things that link to `kronmult_cuda` did, that would be `INTERFACE`. `PUBLIC` would mean both `kronmult_cuda` and things linking to `kronmult_cuda` need those symbols.

bmcdanie replied:
ah but you're talking about `INTERFACE` as it relates to add library

bmcdanie replied:
so this is different, and yes because it's a header only library, we could use that.

hbrunie replied:
Ok. If you look at line 223 of CMakeLists.txt in this PR, `PRIVATE` is indeed used.

bmcdanie reviewed (changes requested):
Hugo - looks good, thank you! if you could help me understand a bit some of the comments, and whether this setup will always pull latest commit from master as written, we can get this in.

hbrunie committed:
Make comment clearer.

TODO: link to kronmult as interface library
I read something about it here: https://foonathan.net/2016/07/cmake-dependency-handling/

bmcdanie reviewed (approved):
thanks.
